### ISSUE TITLE
New Security Interface

### ISSUE THREAD

**Issue by milleruntime** [apache/accumulo#1221]
Design for a cleaner internal security interface:
```java
public interface Security {

  /**
   * Only called when Accumulo is initiazed.  This will initialize the security for Accumulo.
   *
   * @param configuration SiteConfiguration server configuration
   */
  void initialize(SiteConfiguration configuration);

  /**
   * Return a security operation object.
   *
   * @return SecurityOperation Utility class for performing security operations.
   */
  SecurityOperation operation();
}
```

The idea is to separate the security plugin implementation (KerberosAuthenticator, KerberosAuthorizer, ZKAuthorizer, etc), from the ServerContext and Initialization for Accumulo.  Currently, we have bootstrapping security issues between ServerContext and Initialize.  Initialize is a one time action but part way through it creates a ServerContext object and then initializes security.  I think ServerContext should be separated from the Initialize procedure.  This would simplify a lot of things since ServerContext would then always have a security state initialized, which is the typical behavior of a server.

**Comment by ctubbsii** [comment]
The `SecurityOperation` class is currently an internal only class. If this interface were to represent something like we have today, it should look something more like:

```java
public interface Security {
  void initialize(SiteConfiguration configuration);
  Authenticator authenticator();
  Authorizor authorizor();
  PolicyChecker policyChecker(); // or PermissionsHandler
}
```

I don't think we need the second layer interface, `SecurityOperation` if we're going to re-design the pluggable security.

**Comment by milleruntime** [comment]
I like having this in SPI under a security package: ```org.apache.accumulo.core.spi.security```

Then use ```org.apache.accumulo.core.spi.common.ServiceEnvironment``` as a parameter like so:

```java
public interface SecurityModule {
/**
   * Only called when Accumulo is initiazed.  This will initialize the security for Accumulo.
   *
   * @param configuration SiteConfiguration server configuration
   */
  void initialize(ServiceEnvironment env);

  Authenticator authenticator();
  PolicyChecker policyChecker(); // or PermissionsHandler

}
```

I am not convinced we need an ```Authenticator``` and an ```Authorizor```.  If we initialize all the security and server side user creds in the initialize method then we can just handle all the authentication stuff with the Authenticator.  Or just have one thing checking permissions and authentication.  There seems to be a lot of duplicate code across all three of our current pluggable interfaces.  There are some methods like: `dropUser, cleanUser and initUser` which will happen after Intialization, where there might be some overlap.

**Comment by ctubbsii** [comment]
Yes, definitely, anything in the category of "update security because of something that happened in Accumulo" (like deleting a table) should be put in one place. Implementers can initiate multiple actions as a result of the call, but the API for the module should only reflect one call.

**Comment by milleruntime** [comment]
I compiled a list of secured actions to check against a user (I think I have them all, please let me know if I am missing any).  This could replace all the crazy cross checking of different permissions.  The new interface would have this method:
```java
  /**
   * Check if user can perform the action.
   */
  boolean check(String user, Action action);
```
Then the Action enum would have:
```
NAMESPACE_CREATE
NAMESPACE_DROP
NAMESPACE_PROPERTY_CHANGE
NAMESPACE_RENAME
TABLE_BULK_IMPORT
TABLE_CANCEL_COMPACT
TABLE_CLONE
TABLE_COMPACT
TABLE_CONDITIONAL_UPDATE
TABLE_CREATE
TABLE_DROP
TABLE_EXPORT
TABLE_FLUSH
TABLE_IMPORT
TABLE_MERGE
TABLE_OFFLINE
TABLE_ONLINE
TABLE_PROPERTY_CHANGE
TABLE_RANGE_DELETION
TABLE_RENAME
TABLE_SCAN
TABLE_SPLIT
TABLE_SUMMARIZE
TABLE_WRITE
USER_AUTHS_CHANGE
USER_CREATE
USER_DROP
USER_PASSWORD_CHANGE
```
I don't see why we would need to implement java objects or use abstract classes if all we want is just the action name and user.  The business logic would still be in the same place... either as a FATE operation or client operation but that code would call something like `security.check(user, action)`

Then I THINK we only need `initialize(rootuser, token)` and `authenticate(user, token)` methods on the security interface for the rest of the code.

**Comment by ctubbsii** [comment]
I would argue in favor of small classes rather than an enum, for the following reasons:

1. It allows us to organize actions in a type hierarchy with abstract base classes to group them together (not possible with Enums).
2. It allows users to define their own actions in their own code if they want to.
3. It makes our code a bit more extensible if we want to add more actions ourselves.

Example:

```java
interface Action {
  String description();
}

public abstract class TableAction {
  // note these are inner-classes to keep things organized
  public static class Split extends TableAction {
    // ...
  }
  public static class Scan extends TableAction {
    // ...
  }
  // ...
}

```

**Comment by milleruntime** [comment]
It would be good to incorporate ideas in this design that @ctubbsii had previously documented here:
https://issues.apache.org/jira/browse/ACCUMULO-1632

**Comment by milleruntime** [comment]
I have some ideas that I started to implement in a draft PR: https://github.com/apache/accumulo/pull/1736

I am having trouble deciding what the ```User``` class should be and how it should work with the Accumulo internals.  I didn't want to go as far as replace the client methods in [SecurityOperation](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/client/admin/SecurityOperations.java) but the code works well passing ```credentials.getPrincipal()``` to the same methods.  It would be easier to just pass a String for the principal and then we could keep using [Credentials](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/clientImpl/Credentials.java) and [SystemCredentials](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/security/SystemCredentials.java), which extends it.
